### DISCUSSION TITLE
atomWithStorage not working in ReactNative

### DISCUSSION THREAD

**Discussion by ahdam** [pmndrs/jotai#2421] [Bug report]
Hello,

I'm facing issues with atomWithStorage on ReactNative.
Current screen setup:
StackScreen1 - no children
StackScreen2 - has a **TabNavigator** with 3 tabs
   TabScreen1
   TabScreen2
   TabScreen4

I want to save the last accessed tab in **AsyncStorage**.
The code below works when navigating from **StackScreen2** to **StackScreen1** and back to **StackScreen2**. In this case, the value is stored correctly in the atom, and I can read it fine, and set it to the initialRouteName param of the TabNavigator
The problem is **every time I refresh the app, the value from the atom is gone** (by this I mean it reverts to the default one: {value: "Tab 1"}).

Has anyone encountered this before?

This is my setup:

```
// atom creation
import {atomWithStorage, createJSONStorage} from 'jotai/utils';
import AsyncStorage from '@react-native-async-storage/async-storage';

const storage = createJSONStorage(() => AsyncStorage);
export const lastAccessedTab = atomWithStorage(
  'last-tab',
  {value: 'Tab 1'},
  storage,
);
```

```
// update atom
const TabScreen2 = (): React.JSX.Element => {
  const isFocused = useIsFocused();
  const setLastScreen = useSetAtom(lastAccessedTab);

  useEffect(() => {
    if (isFocused) {
      setLastScreen({value: 'Tab 2'});
    }
  }, [isFocused]);

  return (
    <SafeAreaView>
      <Section title="Tab Screen 2">some content</Section>
    </SafeAreaView>
  );
};
```

```
// read from atom
const StackScreen2 = (): React.JSX.Element => {
  const Tab = createMaterialBottomTabNavigator();
  const lastUsedTab = useAtomValue(lastAccessedTab) as any;

  useEffect(() => {
    console.log('lastUsedTab', lastUsedTab);
  }, [lastUsedTab]);

  return (
    <Tab.Navigator initialRouteName={lastUsedTab.value}>
      <Tab.Screen name="Tab 1" component={TabScreen1} />
      <Tab.Screen name="Tab 2" component={TabScreen2} />
      <Tab.Screen name="Tab 3" component={TabScreen3} />
    </Tab.Navigator>
  );
};
```

**Comment by ahdam** [comment]
I found the solution. 
I needed to add the **getOnInit** option.
Without this option, the whole atomWithStorage for ReactNative doesn't work as expected.
I would suggest adding this small thing to the documentation.
Or, even better, make this option default to "true" since only by enabling it this atomWithStorage works as expected.

**Reply by dai-shi** [reply]
It's addressed in https://github.com/pmndrs/jotai/pull/2391.
Do you have a better idea to improve the docs?

**Reply by ahdam** [reply]
Hey @dai-shi, thanks for the reply.

Sorry, it was my mistake, I was looking at the [ReactNative docs](https://jotai.org/docs/utilities/storage)
I read the relevant docs and I think it's perfectly explained there.

I would however suggest that the default value of this flag is **true**. 
Here's why: the reason is that this flag is used for atomWithStorage, which creates an atom with a value that persists, so it makes more sense that the default value of the flag helps with that persistence. If someone wants to deviate from the main function of atomWithStorage, they should change this flag.

Thanks again for taking the time to respond.

**Reply by dai-shi** [reply]
As wrote in the other issue, getOnInit was the default behavior in v1. And, in v2, we flipped the flag based on feedback. Basically, we can't fulfill two different requirements with a single util. We could separate utils, but having an option would be better.

Please feel free to send a PR to improve React Native docs.

**Reply by ahdam** [reply]
I didn't know about the v1 implementation and those discussions.
I double-checked the docs and now, in the light of morning with a fresh coffee in hand, they look good. I see now I skipped the top part of the docs (where the option was listed) and went straight to the ReactNative part.
All is good here, sorry for wasting your time.

**Reply by johannbuscail** [reply]
Hello @dai-shi, even after adding getOnInit, the atom in the storage is gone and everything is reset. Is there any other reason this wouldn't work. Here is my function:

```tsx
const storage = Platform.OS === 'web' ? createJSONStorage() : createJSONStorage(() => AsyncStorage);
export const isOnboardedAtom = atomWithStorage(
	'isOnboardedAtom',
	false,
	storage,
	{
		getOnInit: true,
	}
);

```

**Reply by dai-shi** [reply]
It looks good to me. Can you reproduce it with web?

**Reply by johannbuscail** [reply]
Yes it also resets on the web

**Reply by dai-shi** [reply]
Can you then create a very minimal reproduction with https://stackblitz.com ?

**Reply by johannbuscail** [reply]
All solved ! It actually was an error with Expo that needed to be rebuilt. Thanks!

**Comment by adelpro** [comment]
```const storage = Platform.OS === 'web' ? createJSONStorage() : createJSONStorage(() => AsyncStorage);
export const isOnboardedAtom = atomWithStorage(
	'isOnboardedAtom',
	false,
	storage,
	{
		getOnInit: true,
	}
);
```
How to add types here? thanks

**Reply by dai-shi** [reply]
https://github.com/pmndrs/jotai/blob/7c8ea63510ae0a47da6de092c9dcf20415e5cec9/src/vanilla/utils/atomWithStorage.ts#L211-L216

Ideally `storage` could be typed. Otherwise you can explicitly pass `Value` type.
